### PR TITLE
[#1144] Send.Form screen rework

### LINKED ISSUE #340: Design: Display Errors When Sending Zec
The send ZEC screen should provide good feedback to the user about errors on the form input.  Some examples of errors:

- [x] Amount too small (cannot cover fee)
- [x] Amount too large (exceeds current balance)
- [x] Invalid amount
- [x] Bad address
- [x] Insufficient funds
- [x] Invalid address

### LINKED ISSUE #810: Disable Memo UI field in case of Transparent address
## Is your feature request related to a problem? Please describe.
We'd like to improve our Send screen UI with mechanisms of hiding/disabling Memo UI field in case of Transparent address detected as receiver. 

## Describe the solution you'd like
Keep the Memo field in place and disable/enable it while receiver address validated and recognised its type. Show the Memo field in the red error state and add there an error message below it.  Also don't let user to hit the Send button (disable it) unless all characters from the Memo field removed.

Validation of the receiver address could be done with simply `t` starting character checking for now.

## Alternatives you've considered
Alternatively we could consider usage of `isValid...` methods from the SDK. 

These methods maybe not need to be suspend
Maybe we could move these addresses validations from `Synchronizer` to Wallet Address object, and we need then to keep the suspend https://github.com/zcash/zcash-android-wallet-sdk/issues/342. We have an issue to make the wallet address validation more type safe.  Ultimately, we probably need one more companion function on WalletAddress in `sdk-incubator-lib` that takes in a string (for address and returns the correct sealed class type).

### LINKED ISSUE #826: SendArgumentsWrapper object properties validation
## Is your feature request related to a problem? Please describe.
We'd like to have the `SendArgumentsWrapper` class properties validated, and thus create a type-safe model class.

We have a concern that the `SendArgumentsWrapper` is not type-safe, so the fields contained within it may not necessarily be valid. This would allow bypassing the live validation we have as the user types, meaning incorrect values could get pre-filled into the Send.Form screen text fields.

## Describe the solution you'd like
We're not sure about the most suitable solution here, but it could be e.g.
- Adding type safety to the `SendArgumentsWrapper` so that we skip scanning bad QR codes
- Adding validation to Send.Form screen before we use the values from `SendArgumentsWrapper`
- Adding validation before we insert values to the `SendArgumentsWrapper` model data class

We realize that the arguments we pass through the Navigation library are all strings, so that complicates type safety a bit. We could validate before serializing to strings, then assume they're safe upon deserializing since it's an internal API.

### LINKED ISSUE #1157: Send Form memo length counter

### LINKED ISSUE #1158: Send Form memo field availability

### LINKED ISSUE #1253: Change fee label to say "typical fee < 0.001"
We will be enabling ZIP 317 fees for 1.0; this ensures that if we don't have time to expose the actual fees, users are still being told something true.

### LINKED ISSUE #1254: Ensure that all errors related to transaction creation are shown to users on the send screen
#340 is about doing real time validation of inputs as users fill in the Send screen. This is a simpler issue that could be solved by removing all real time validation and instead relying on the existing error message reporting of send-time errors (though once #1159 is done, those errors would show on the send confirmation screen instead, which is less great but still better than the user not knowing why they can't click the Send button).

### PR DESCRIPTION
**Pull request by HonzaR**
- This follows the new Figma design. It also adds the TextFields values validation and proper UI reactions.
- Closes #340
- Closes #810
- Closes #1157
- Closes #1158
- Closes #1253
- Closes #1254
- Closes #826
- Follow-ups: #1047, #1257

**AFTER CHANGES:** (Note the screen rotation only demonstrates the TextFields proper values persisting rather than a real-world situation, as Zashi has rotation currently locked in the portrait mode)

https://github.com/Electric-Coin-Company/zashi-android/assets/9017390/e15194fc-7ce8-4be4-84ed-6142204677e8

**BEFORE CHANGES:**

> **Note**
>This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] **Self-review** your own code in GitHub's web interface[^1]
- [ ] Add **automated tests** as appropriate
- [ ] Update the [**manual tests**](../blob/main/docs/testing/manual_testing)[^2] as appropriate
- [ ] Check the **code coverage**[^3] report for the automated tests
- [ ] Update **documentation** as appropriate (e.g [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), [**CHANGELOG.md**](../blob/main/CHANGELOG.md), etc.)
- [x] **Run the app** and try the changes
- [x] Pull in the latest changes from the **main** branch and **squash** your commits before assigning a reviewer[^4]

> **Note**
> It is good practice to provide before and after UI  **screenshots** in the description of this PR. This is only applicable for changes that modify the UI.

# Reviewer

- [ ] Check the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) **checklist**
- [ ] Perform an **ad hoc review**[^5]
- [ ] Review the **automated tests**
- [ ] Review the **manual tests**
- [ ] Review the **documentation**, [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), etc. as appropriate
- [ ] **Run the app** and try the changes[^6]

[^1]: _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
[^2]: _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
[^3]: _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed. Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules._
[^4]: _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._
[^5]: _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
[^6]: _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

**Comment by daira** [comment]
I noticed from the video that the memo was not cleared when you pasted in the t-address. I think this is different from Zashi iOS.

File: ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/TextField.kt [hunk 1]
```diff
@@ -31,40 +32,53 @@ fun FormTextField(
...
     textStyle: TextStyle = ZcashTheme.extendedTypography.textFieldValue,
-    label: @Composable (() -> Unit)? = null,
-    leadingIcon: @Composable (() -> Unit)? = null,
-    trailingIcon: @Composable (() -> Unit)? = null,
+    placeholder:
+        @Composable()
+        (() -> Unit)? = null,
+    leadingIcon:
+        @Composable()
+        (() -> Unit)? = null,
```
Comment (LukasKorba) [inline ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/TextField.kt:43]: What leadingIcon does? Seems like it's something from the previous design we had months or even years ago. Do we want to keep it or clean up the code?

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/compose/BalanceWidget.kt [hunk 2]
```diff
@@ -40,19 +40,21 @@ private fun BalanceWidgetPreview() {
...
+                                WalletBalance(
+                                    Zatoshi(1234567891234567),
+                                    Zatoshi(123456789),
+                                    Zatoshi(123)
+                                )
+                        ),
+                    isReferenceToBalances = true,
+                    onReferenceClick = {},
+                    modifier = Modifier
+                )
```
Comment (LukasKorba) [inline ui-lib/src/main/java/co/electriccoin/zcash/ui/common/compose/BalanceWidget.kt:57]: This is more about my curiosity, I checked the raw file and the only change here is that exactly the same code has been wrapped into parentheses (). Is it by accident or it has some meaning?

**Comment by HonzaR** [comment]
> I noticed from the video that the memo was not cleared when you pasted in the t-address. I think this is different from Zashi iOS.

Yes, it's not cleared from the memo field, but the field is disabled, and the message from the memo is not used when the Preview button hits. The pro of this solution is that the message value is not lost when the address is changed to a unified or sapling one. @LukasKorba @true-jared @adjyrichard What do you think?

File: ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/TextField.kt (hunk 1 above)
Comment (HonzaR) [reply ui-design-lib/src/main/java/co/electriccoin/zcash/ui/design/component/TextField.kt:43]: The TextField is a reusable component accessible from our `ui-design-lib` module, which makes it an API with a different use case across the app's UI. Having the leading icon parameter defined as nullable is fine even if not used. The `trailingIcon` parameter is used on the Send.Form for the recipient address TextField, so logically, the API could also provide the `loadingIcon.`

**Review by LukasKorba — APPROVED** [review]
utACK, 2 non-blockers

**Comment by LukasKorba** [comment]
> > I noticed from the video that the memo was not cleared when you pasted in the t-address. I think this is different from Zashi iOS.
> 
> Yes, it's not cleared from the memo field, but the field is disabled, and the message from the memo is not used when the Preview button hits. The pro of this solution is that the message value is not lost when the address is changed to a unified or sapling one. @LukasKorba @true-jared @adjyrichard What do you think?

iOS is clearing it as it was required. If Android is behaving differently I think it's better if it's discussed prior PR.

**Comment by LukasKorba** [comment]
I noticed in the video that memo field's height is expanding, it was supposed to be fixed size.

File: ui-lib/src/main/java/co/electriccoin/zcash/ui/common/compose/BalanceWidget.kt (hunk 2 above)
Comment (HonzaR) [reply ui-lib/src/main/java/co/electriccoin/zcash/ui/common/compose/BalanceWidget.kt:57]: Having it in parentheses is IMHO not necessary and was caused by Ktlint auto-formtting.

**Comment by true-jared** [comment]
@HonzaR We have discussed the memo behaviour, and made a decision to clear the message. Regarding the fixed height/expandable height of the input fields, I understand your UX perspective but I 100% agree that making any changes to the requirements/designs while working on an issue should be raised prior to making a PR and definitely before merging any changes you have made. I don't want to hold you back and make you rework anything right now, but please respect the process going forward. We can discuss/track any improvements you would like to add, but to meet the release deadline, we need to focus on the hard functional requirements first, and not spend time on nice-to-haves.

**Comment by true-jared** [comment]
Filed here: #1269 @HonzaR .

**Pull request merged**